File: Algorithms/src/2657. Find the Prefix Common Array of Two Arrays/Code/solution.go
func findThePrefixCommonArray(A []int, B []int) []int {
    
    n := len(A)
    
    // Frequency array to track appearances
    freq := make([]int, n+1)
    
    // Result array
    ans := make([]int, n)
    
    // Stores count of common elements
    common := 0
    
    for i := 0; i < n; i++ {
        
        // Add current element from A
        freq[A[i]]++
        
        // If frequency becomes 2,
        // number appeared in both arrays
        if freq[A[i]] == 2 {
            common++
        }
        
        // Add current element from B
        freq[B[i]]++
        
        // Same check for B
        if freq[B[i]] == 2 {
            common++
        }
        
        // Store current prefix answer
        ans[i] = common
    }
    
    return ans
}

File: Algorithms/src/2657. Find the Prefix Common Array of Two Arrays/Code/solution.cpp
class Solution
{
public:
    vector<int> findThePrefixCommonArray(vector<int> &A, vector<int> &B)
    {

        int n = A.size();

        // Frequency array to track how many times
        // each number has appeared so far
        vector<int> freq(n + 1, 0);

        // Result array
        vector<int> ans(n);

        // Stores current count of common numbers
        int common = 0;

        for (int i = 0; i < n; i++)
        {

            // Add current element from A
            freq[A[i]]++;

            // If frequency becomes 2,
            // it means number appeared in both arrays
            if (freq[A[i]] == 2)
            {
                common++;
            }

            // Add current element from B
            freq[B[i]]++;

            // Same check for B
            if (freq[B[i]] == 2)
            {
                common++;
            }

            // Store answer for current prefix
            ans[i] = common;
        }

        return ans;
    }
};